pkg/proxy/iptables/proxier.go

```diff
@@ -1107,16 +1110,22 @@ func (proxier *Proxier) syncProxyRules() {
 				"--dport", strconv.Itoa(svcInfo.Port()),
 			)
 			if proxier.masqueradeAll {
-				utilproxy.WriteRuleLine(proxier.natRules, string(svcChain), append(args, "-j", string(KubeMarkMasqChain))...)
+				args := prepend(args, "-A", string(svcChain))
+				args = append(args, "-j", string(KubeMarkMasqChain))
+				utilproxy.WriteLine(proxier.natRules, args...)
 			} else if proxier.localDetector.IsImplemented() {
 				// This masquerades off-cluster traffic to a service VIP.  The idea
 				// is that you can establish a static route for your Service range,
 				// routing to any node, and that node will bridge into the Service
 				// for you.  Since that might bounce off-node, we masquerade here.
 				// If/when we support "Local" policy for VIPs, we should update this.
-				utilproxy.WriteRuleLine(proxier.natRules, string(svcChain), proxier.localDetector.JumpIfNotLocal(args, string(KubeMarkMasqChain))...)
+				args := prepend(args, "-A", string(svcChain))
+				args = proxier.localDetector.JumpIfNotLocal(args, string(KubeMarkMasqChain))
+				utilproxy.WriteLine(proxier.natRules, args...)
 			}
-			utilproxy.WriteRuleLine(proxier.natRules, string(kubeServicesChain), append(args, "-j", string(svcChain))...)
+			args = prepend(args, "-A", string(kubeServicesChain))
```

Comment:
so this breaks the "not reallocating `args` thing we had before". `args` was initialized as

```
	args := make([]string, 64)
```

at the top of `syncProxyRules`, before the start of the `range proxier.serviceMap` loop, and then each time we started a new arg list, it would do `args = args[:0]` to truncate it while keeping the underlying storage.

but a single `args = prepend(...)` breaks that for the entire rest of the current `syncProxyRules`. So we will now be doing A LOT more memory allocations than we had been doing before.

Reply:
great catch
```
package main

import (
	"testing"
)

func BenchmarkPrependAlloc(b *testing.B) {
	a := make([]int, b.N)

	for i := 0; i < b.N; i++ {
		a = prepend(a, i)
	}
}

func BenchmarkAlloc(b *testing.B) {
	a := make([]int, b.N)

	for i := 0; i < b.N; i++ {
		a = append(a, i)
	}
}

func prepend(sl []int, args ...int) []int {
	return append(args, sl...)
}
```
```
goos: linux
goarch: amd64
cpu: Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
BenchmarkPrependAlloc-12    	  112240	    171318 ns/op
BenchmarkAlloc-12           	164673328	         6.298 ns/op
PASS
```

Reply:
We can either a) change prepend to allocate the full capacity of args (so it allocs but only once); b) change prepend to shuffle the elements down and insert into the same slice; c) change Write to take a [][]string (scatter-gather style).  The truth is that every time we do 

If we want to take this seriously, we should explore how to do it better in all cases.

Reply:
> change Write to take a [][]string (scatter-gather style). The truth is that every time we do

Your sentence got cut off, but whatever you were going to say, that idea sounds really good...

```
matchClusterIP := []string{
	"-m", protocol, "-p", protocol,
	"-d", utilproxy.ToCIDR(svcInfo.ClusterIP()),
	"--dport", strconv.Itoa(svcInfo.Port()),
}

if hasEndpoints {
	comment := utilproxy.Comment("%s cluster IP"`, svcNameString)
	if proxier.masqueradeAll {
		proxier.natRules.Write(
			[]string{"-A", string(svcChain)},
			comment,
			matchClusterIP,
			[]string{"-j", string(KubeMarkMasqChain)},
		)
	} else if proxier.localDetector.Implemented() {
		// ...
		proxier.natRules.Write(
			[]string{"-A", string(svcChain)},
			comment,
			proxier.localDetector.JumpIfNotLocal(KubeMarkMasqChain),
		)
	}
	proxier.natRules.Write(
		[]string{"-A", string(kubeServicesChain)},
		comment,
		matchClusterIP,
		[]string{"-j", string(svcChain)},
	)
} else {
	// No endpoints.
	proxier.filterRules.Write(
		[]string{"-A", string(kubeServicesChain)},
		utilproxy.Comment("%s has no endpoints", svcNameString),
		matchClusterIP,
		[]string{"-j", "REJECT"},
	)
}
```

(`utilproxy.Comment` is left as an exercise to the reader). Maybe even

```
func addToChain(chain utiliptables.Chain) []string {
	return []string{"-A", string(chain)}
}

func jumpToChain(chain utiliptables.Chain) []string{
	return []string{"-j", string(chain)}
}
```

Reply:
Or even have `Write(args ...interface{})` and let you pass an arbitrary combination of `string`, `[]string`, and `utiliptables.Chain`...

```
	proxier.filterRules.Write(
		"-A", kubeServicesChain,
		utilproxy.Comment("%s has no endpoints", svcNameString),
		matchClusterIP,
		"-j", "REJECT",
	)
```

Reply:
I'm trying to craft a representative benchmark.

Reply:
I'm trying to see if I can get something with https://gist.github.com/aojea/04999dbda0f9f109c19513363b91269a

Reply:
Benchmark: https://gist.github.com/thockin/8fc599cc4adc7aba72a367057704a798

```
$ go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: foo
cpu: Intel(R) Xeon(R) W-2135 CPU @ 3.70GHz
BenchmarkOneAppend-6              	12748252	        94.44 ns/op	      48 B/op	       1 allocs/op
BenchmarkOneVariadic-6            	12794826	        93.48 ns/op	      48 B/op	       1 allocs/op
BenchmarkOnePrepend-6             	 6587660	       182.3 ns/op	     176 B/op	       2 allocs/op
BenchmarkOnePrependInPlace-6      	11071024	       108.2 ns/op	      48 B/op	       1 allocs/op
BenchmarkOneSG-6                  	11634345	       103.1 ns/op	      48 B/op	       1 allocs/op
BenchmarkAppend-6                 	 3653259	       290.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrepend-6                	  378285	      3120 ns/op	    6464 B/op	      32 allocs/op
BenchmarkPrependInPlaceManual-6   	 1499826	       706.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrependInPlaceCopy-6     	 2113896	       533.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkScatterGatherFlat-6      	  997440	      1113 ns/op	     608 B/op	      17 allocs/op
BenchmarkScatterGatherDeep-6      	  834326	      1382 ns/op	    1120 B/op	      33 allocs/op
```

So a write() call with an append is the same as a variadic.

Prepend, as written, allocates and is much slower.

Changing write() to `...[]string` is fast.





3:31 PM
in-place prepend is fast, but destructive (so violates POLS) (edited) 
3:32 PM
scatter-gather is fast, combining it with prepend is slower than prepend

Reply:
Adding a string-or-slice mode to write seems plausible, and makes the net diff smaller AND opens further opportunities. 

I want to be careful not to make this code any less readable than it already is.  Let me try a couple things.

Reply:
Updated.  I'll take it apart:

```
BenchmarkOneAppend-6                   	12719592	        95.78 ns/op	      48 B/op	       1 allocs/op
BenchmarkOneVariadic-6                 	13332460	        90.14 ns/op	      48 B/op	       1 allocs/op
BenchmarkOnePrepend-6                  	 6574177	       183.5 ns/op	     176 B/op	       2 allocs/op
BenchmarkOnePrependInPlace-6           	11058524	       108.8 ns/op	      48 B/op	       1 allocs/op
BenchmarkOneSG-6                       	11836407	       100.9 ns/op	      48 B/op	       1 allocs/op
BenchmarkOneIntf-6                     	12110347	        98.34 ns/op	      48 B/op	       1 allocs/op
```

No penalty for variadic calls.  Prepend is bad.  In-place prepend is not bad but surprising.  Scatter-gather and using an interface both are within spitting distance of no-op.

Run these under more load to see amortize overhead out:

```
BenchmarkAppend-6                      	 3696194	       286.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrepend-6                     	  394273	      2944 ns/op	    6464 B/op	      32 allocs/op
BenchmarkPrependInPlaceManual-6        	 1438062	       747.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkPrependInPlaceCopy-6          	 2344239	       519.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkScatterGatherFlat-6           	 1000000	      1064 ns/op	     608 B/op	      17 allocs/op
BenchmarkScatterGatherDeep-6           	  897902	      1352 ns/op	    1120 B/op	      33 allocs/op
```

Append doesn't actually work, but it sets the bar.  Prepend is 10x slower.  In-place prepend is only 2x slower.  Scatter-gather turns into a different form of the prepend problem.  These results are not very interesting.

A more representative test (look at the gist for code):

```
BenchmarkReprSGAsFewAsPossible-6       	11684551	        92.63 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprSGMany-6                  	 3608090	       333.5 ns/op	     288 B/op	       5 allocs/op
BenchmarkReprSGManyInline-6            	10851805	       100.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfAsFewAsPossible-6   	11803284	        91.16 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfMany-6              	 5385872	       235.8 ns/op	     160 B/op	       3 allocs/op
BenchmarkReprIntfManyInline-6        	10768394	        98.16 ns/op	       0 B/op	       0 allocs/op
```

These show that both SG and intf are viable, and depending how we use them can perform well.  Some code readability considerations...

To explore code reading:

```
BenchmarkReprClauseAsFewAsPossible-6   	11474544	        91.50 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprClauseMany-6              	 4877307	       244.6 ns/op	     128 B/op	       4 allocs/op
BenchmarkReprClauseManyInline-6        	 6517003	       182.9 ns/op	      72 B/op	       2 allocs/op
```

This adds a function-type argument.  I like the code but perf leaves something to be desired.

```
BenchmarkReprFuncAsFewAsPossible-6     	11681811	        91.60 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprFuncAsFewAsPossible2-6    	10203375	       103.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprFuncMany-6                	 5535034	       216.0 ns/op	     160 B/op	       3 allocs/op
BenchmarkReprFuncManyInline-6          	10537470	       101.1 ns/op	       0 B/op	       0 allocs/op
```

This is more or less Intf with function helpers.  I am not sure it makes the code any clearer.  

Take a look and tell me what you think

Reply:
One more:

We could define a "WordList" and "WordSlice" type, which manage an offset slice underneath so prepends are free. Code would look like:

```
wl := NewWordList(64, 8) // capacity and zero-offset
// ... later
sl := a.Slice(0, 0)  // a view into wl
sl = sl.Append("-m", protocol, "-p", protocol)
sl = sl.Append("-d", ip, "--dport", port)
sl = sl.Append("-j", "FOOBAR")
proxier.natRules.write(sl.Prepend("-X", "CHAIN").Slice())
```

We could user helper funcs if we think it is cleaner:

```
sl = sl.Append(protocolFunc(protocol)...)
sl = sl.Append(destFunc(ip, port)...)
sl = sl.Append(jumpFunc("FOOBAR")...)
proxier.natRules.write(sl.Prepend(appendFunc("CHAIN")...).Slice())
```

We could maybe drop the `...`.

We could maybe even get more DSLish, maybe?

I wrote a simple form of the code, it's only 80 LOC.  Scatter-gather and Intf are simpler (less LOC) but maybe this is cleaner?  Not sure.

Reply:
> These show that both SG and intf are viable, and depending how we use them can perform well. Some code readability considerations...

`BenchmarkReprIntfManyInline` is using arrays that it doesn't need to use though:

```
		writeIntf(&buf,
			"-A", "CHAIN",
			[]string{"-m", protocol, "-p", protocol},
			[]string{"-d", ip, "--dport", port},
			[]string{"-j", "FOOBAR"})
```

This could just be

```
		writeIntf(&buf,
			"-A", "CHAIN",
			"-m", protocol, "-p", protocol,
			"-d", ip, "--dport", port,
			"-j", "FOOBAR")
```

which doesn't seem unreadable to me at all.

Of course, it also doesn't test the `interface{}` functionality. But we could do:

```
		matchService := []string{
			"-m", protocol, "-p", protocol,
			"-d", ip, "--dport", port,
		}
		writeIntf(&buf,
			"-A", "CHAIN",
			matchService,
			"-j", "FOOBAR")
```

which again seems readable to me.

FTR

```
BenchmarkReprIntfAsFewAsPossible-12            	15124519	        80.31 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfManyInlinePreArray-12         	14044212	        82.45 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfManyInlineNoExtraArrays-12    	14103712	        84.79 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfManyInlineNonRecursive-12     	13345648	        88.40 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfManyInline-12                 	12820171	        90.49 ns/op	       0 B/op	       0 allocs/op
BenchmarkReprIntfMany-12                       	 2739234	       437.9 ns/op	     424 B/op	       8 allocs/op
```

(`PreArray` is the one with `matchService`, `NoExtraArrays` is the one with all the arrays replaced with individual strings. `NonRecursive` was another experiment, with making `writeIntf` loop over `[]string{}` arguments itself rather than doing `write(buf, x...)`.)

https://gist.github.com/danwinship/b5524b557dc8ab8b544ab25f0df9ecf7

> We could define a "WordList" and "WordSlice" type, which manage an offset slice underneath so prepends are free.

I think we can just avoid needing to do prepends

> We could maybe even get more DSLish, maybe?

Yeah, I was wondering about that when I suggested `utilproxy.Comment`, `jumpToChain` etc... I'm not sure if it's a good idea, because we're not going to be able to completely abstract away iptables syntax, so we'd probably just end up making it so that people need to learn both iptables syntax _and_ our DSL if they want to hack on the code. It may be better to just keep all the iptables syntax explicit, but just make it easier to efficiently reuse bits and pieces of it as needed.

Reply:
> using arrays that it doesn't need to use

I was just trying to push the pattern and see where it broke down.  Using many slices allows the proliferation of "builder" functions which each return a `[]string`.

> It may be better to just keep all the iptables syntax explicit, but just make it easier to efficiently reuse bits and pieces of it as needed.

+1

The main downside of `interface{}` is that it changes a compile time error (passing the wrong type) into a runtime error.

Shall we agree that everything but scatter-gather and interface models are ruled out?